### PR TITLE
Rework compare operator

### PR DESCRIPTION
**Pull request by qingxinhome**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3440

## What this PR does / why we need it:
Using CGO to reconstruct the comparison operation

### CODE REVIEW

**Review by fengttt — COMMENTED** [review]
Does equal, ge, gt, lt, le have to have their own impl in the vectorize package?   Aren't they doing same thing?

**Comment by qingxinhome** [comment]
> Does equal, ge, gt, lt, le have to have their own impl in the vectorize package? Aren't they doing same thing?

I understand that they are indeed similar operations. I will make the comparison operation another layer of abstraction and remove the redundant code, but I think they are still vectorized operations

File: cgo/compare.c [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    } else if (type == Type_UINT8) {
+        MO_COMPARE_T(COMPARE_NE, uint8_t);
+    } else if (type == Type_UINT16) {
+        MO_COMPARE_T(COMPARE_NE, uint16_t);
+    } else if (type == Type_UINT32) {
+        MO_COMPARE_T(COMPARE_NE, uint32_t);
+    } else if (type == Type_UINT64) {
+        MO_COMPARE_T(COMPARE_NE, uint64_t);
+    } else if (type == Type_FLOAT32) {
+        MO_COMPARE_T(COMPARE_NE, float);
```
Comment (daviszhen) [inline cgo/compare.c:207]: How to handle the approximate equal on the float and double type?

File: cgo/compare.c [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+/*
+ * less equal operator (<=)
+ */
+#define COMPARE_LE(TGT, A, B)                                 \
+    TGT = ((A) <= (B))
+
+
+/*
+ * bool compare operator
+ */
```
Comment (daviszhen) [inline cgo/compare.c:84]: Are you sure that the value of the bool in Golang is same as the one in CLang?
So, the  '==' and  '!=' for the equal or not equal operators of the bool needs to be considered.

File: cgo/compare.c [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+#define COMPARE_BOOL_EQ(TGT, A, B)                            \
+    TGT = ((A) == (B))
+
+#define COMPARE_BOOL_NE(TGT, A, B)                            \
+    TGT = ((A) != (B))
+
+#define COMPARE_BOOL_LE(TGT, A, B)                            \
+    TGT = ((!A) || (B))
```
Comment (daviszhen) [inline cgo/compare.c:92]: Change "(!A)" to "!(A)". Same as follow.

**Review by daviszhen — CHANGES_REQUESTED** [review]

File: cgo/compare.c (hunk 1 above)
Comment (qingxinhome) [reply cgo/compare.c:207]: Float and double only compare the same types, and can be compared directly. so there is no case of  approximate equal.

File: cgo/compare.c (hunk 3 above)
Comment (qingxinhome) [reply cgo/compare.c:92]: OK

File: cgo/compare.c (hunk 2 above)
Comment (qingxinhome) [reply cgo/compare.c:84]: OK， I use the C language style to compare bool types

**Review by ouyuanning — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Pull request merged**